### PR TITLE
DNM: ceph: default to clang for crimson builds

### PR DESCRIPTION
**Pull request by athanatos**
<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Review by tchaikov — APPROVED** [review]
lgtm. would be great to see it built on https://shaman.ceph.com as well.

**Comment by athanatos** [comment]
Yeah, I pushed it to the ci, but it hasn't quite gotten to it yet.

**Comment by tchaikov** [comment]
crimson flavor failed to build
```
-- Performing Test HAVE_ATTR_SYMVER - Failed
CMake Error at cmake/modules/CephChecks.cmake:161 (message):
  please pass -flto-partition=none as part of CXXFLAGS
Call Stack (most recent call first):
  CMakeLists.txt:122 (include)
```
see https://shaman.ceph.com/builds/ceph/wip-sjust-crimson-clang/cb6c9d5d74d2e42bbeb1bd3ff2e3b1fde6900a74/

Clang does not accept the suggested option. see https://clang.llvm.org/docs/UsersManual.html . it's very likely that the issue of https://tracker.ceph.com/issues/40060 does not apply to Clang. or we need to reevaluate the status quo of GCC. either way, we need to revise the check in CMake.

**Comment by athanatos** [comment]
Seems to be specific to librados -- the crimson build shouldn't actually affect that.

**Comment by athanatos** [comment]
Repushed with an exception for clang, let's see how that goes.  Looks like we don't actually need the ceph-build change -- the centos9 builder uses the pkg.ceph.crimson profile.

File: cmake/modules/CephChecks.cmake [hunk 1]
```diff
@@ -155,9 +155,9 @@ __attribute__((__symver__ (\"func@v2\"))) void func_v2() {};
 
 int main() {}"
   HAVE_ATTR_SYMVER)
-  if(NOT HAVE_ATTR_SYMVER)
+  if(NOT HAVE_ATTR_SYMVER AND CMAKE_COMPILER_IS_GNUCXX)
```
Comment (tchaikov) [inline cmake/modules/CephChecks.cmake:158]: quote from CMake document:

> This variable is deprecated. Use [CMAKE_CXX_COMPILER_ID](dfile:///Users/kefu/Library/Application%20Support/Dash/DocSets/CMake/CMake.docset/Contents/Resources/Documents/cmake.org/cmake/help/v3.28/variable/CMAKE_LANG_COMPILER_ID.html#variable:CMAKE_%3CLANG%3E_COMPILER_ID) instead

so we could use:

```cmake
if(CMAKE_CXX_COMPILER_ID STREQUAL GNU AND NOT HAVE_ATTR_SYMVER)
```

**Comment by athanatos** [comment]
https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos9,DIST=centos9,MACHINE_SIZE=gigantic/77198//consoleFull

Hmm, it's failing to find libatomic:

```
-- Performing Test HAVE_CXX11_ATOMIC
-- Performing Test HAVE_CXX11_ATOMIC - Failed
-- Performing Test HAVE_LIBATOMIC
-- Performing Test HAVE_LIBATOMIC - Failed
CMake Error at cmake/modules/CheckCxxAtomic.cmake:67 (message):
  Host compiler /usr/bin/clang++ requires libatomic, but it is not found
Call Stack (most recent call first):
  CMakeLists.txt:148 (include)
```

Slightly earlier:

```
Package libatomic-11.4.1-3.el9.x86_64 is already installed.
```

**Comment by tchaikov** [comment]
i am able to build with clang on fc39 though.
```console
$ grep -w CMAKE_CXX_COMPILER build/CMakeCache.txt
CMAKE_CXX_COMPILER:STRING=/home/kefu/.local/bin/clang++
//ADVANCED property for variable: CMAKE_CXX_COMPILER
CMAKE_CXX_COMPILER-ADVANCED:INTERNAL=1
$ grep ATOMIC build/CMakeCache.txt
//Test HAVE_CXX11_ATOMIC
HAVE_CXX11_ATOMIC:INTERNAL=
//Test HAVE_LIBATOMIC
HAVE_LIBATOMIC:INTERNAL=1
```

**Comment by athanatos** [comment]
Yeah, and myself on centos 9 with the same version of libatomic.  I checked the cmake unit test code as well, works fine.

**Comment by athanatos** [comment]
I'll have to try to get at the output from the failing build.

**Comment by athanatos** [comment]
Alright, I managed to reproduce it in a clean centos 9 container by building doing an rpm build locally.  The error from CMakeConfigureLog.yaml is:

```
/usr/bin/clang++ -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security  -Wp,-D_GLIBCXX_ASSERTIONS
-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protect
ion -fcf-protection  -std=c++11 -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld  CMakeFiles/cmTC_73e5f.dir/src.cxx.o -o
cmTC_73e5f  -latomic
        clang++: warning: argument unused during compilation: '-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1' [-Wunused-command-line-argument]
        clang++: warning: argument unused during compilation: '-specs=/usr/lib/rpm/redhat/redhat-hardened-ld' [-Wunused-command-line-argument]
        /opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/../../../../bin/ld: cannot find -latomic: No such file or directory
```

I reconstructed the test source file and confirmed that it really does fail:

```
/usr/bin/clang++ -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security  -Wp,-D_GLIB
CXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -m64 -march=x86-64-v2 -mtune=generic -fasynchronous-unwind-tables -fstac
k-clash-protection -fcf-protection  -std=c++11 -Wl,-z,relro -Wl,--as-needed  -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld  atomics.cc -ocmTC_73e5f
 -latomic
clang++: warning: optimization flag '-ffat-lto-objects' is not supported [-Wignored-optimization-argument]
clang++: warning: argument unused during compilation: '-specs=/usr/lib/rpm/redhat/redhat-hardened-cc1' [-Wunused-command-line-argument]
clang++: warning: argument unused during compilation: '-specs=/usr/lib/rpm/redhat/redhat-hardened-ld' [-Wunused-command-line-argument]
/opt/rh/gcc-toolset-13/root/usr/lib/gcc/x86_64-redhat-linux/13/../../../../bin/ld: cannot find -latomic: No such file or directory
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
```

```sudo dnf install gcc-toolset-13-libatomic-devel.x86_64``` fixes it.

Something seems to be causing it to use the gcc-toolset-13 path, trying to figure out why.

**Comment by tchaikov** [comment]
clang has the builtin support of gcc-toolset. i do remember it even hardwires to some of the gcc-toolset versions.

**Comment by athanatos** [comment]
Yeah, that appears to be the case, even with
source /opt/rh/gcc-toolset-12/enable
LD_LIBRARY_PATH=/opt/rh/gcc-toolset-12/root/usr/lib64:/opt/rh/gcc-toolset-12/root/usr/lib
PKG_CONFIG_PATH=/opt/rh/gcc-toolset-12/root/usr/lib64/pkgconfig

it still looks for the 13 one.

**Comment by athanatos** [comment]
Checking what centos 8 does

**Comment by athanatos** [comment]
Both seem to want -13, *shrug*

```
diff --git a/ceph.spec.in b/ceph.spec.in
index 38bf0b4543c..ece9017cf5f 100644
--- a/ceph.spec.in
+++ b/ceph.spec.in
@@ -354,6 +354,7 @@ BuildRequires:  libasan
 BuildRequires:  protobuf-devel
 BuildRequires:  protobuf-compiler
 BuildRequires:  clang >= 16.0.0
+BuildRequires:  gcc-toolset-13-libatomic-devel
 %if 0%{?rhel} == 8
 BuildRequires:  %{gts_prefix}-annobin
 BuildRequires:  %{gts_prefix}-annobin-plugin-gcc
```

**Comment by tchaikov** [comment]
i created #55886 to address this. it's likely to conflict to your change. if it works, probably we can absorb it in this PR, and close it.

**Comment by tchaikov** [comment]
i just applied the proposed change on top of this PR. it is being built at https://shaman.ceph.com/builds/ceph/wip-pr-55787-kefu/95e03f88094918005a14c2ac719b991c40e656ab/crimson/384274/

File: ceph.spec.in [hunk 2]
```diff
@@ -1463,6 +1464,8 @@ cmake .. \
 %if 0%{with seastar}
     -DWITH_SEASTAR:BOOL=ON \
     -DWITH_JAEGER:BOOL=OFF \
+    -DCMAKE_C_COMPILER=clang \
+    -DCMAKE_CXX_COMPILER=clang++ \
```
Comment (tchaikov) [inline ceph.spec.in:1468]: seems we also need to add `%global toolchain clang`. i am adding it like:
```diff
diff --git a/ceph.spec.in b/ceph.spec.in
index 281db563bc0..262f03c6e28 100644
--- a/ceph.spec.in
+++ b/ceph.spec.in
@@ -1335,6 +1335,9 @@ This package provides a Ceph hardware monitoring agent.
 %if ( 0%{?rhel} && 0%{?rhel} < 9 ) || ( 0%{?suse_version} && 0%{?suse_version} <= 1500 )
 %define _lto_cflags %{nil}
 %endif
+%if 0%{with seastar}
+%global toolchain clang
+$endif

 %if 0%{with cephfs_java}
 # Find jni.h
```
see https://src.fedoraproject.org/rpms/redhat-rpm-config/blob/4637e1bd5512b869bd07c58b7545d2528a9bc4c8/f/macros#_44

**Comment by athanatos** [comment]
> i created #55886 to address this. it's likely to conflict to your change. if it works, probably we can absorb it in this PR, and close it.

Upgrading to gcc-13 also happens to solve my coroutine problem.  I'm hitting enough issues with switching over to clang that at this point I'd be fine with that outcome.

**Comment by tchaikov** [comment]
> i just applied the proposed change on top of this PR. it is being built at https://shaman.ceph.com/builds/ceph/wip-pr-55787-kefu/95e03f88094918005a14c2ac719b991c40e656ab/crimson/384274/

testing at https://shaman.ceph.com/builds/ceph/wip-pr-55825-kefu/80b5a35e74aad49d5c48803310e359473a086d03/default/384342/, which also contains #55890

**Comment by athanatos** [comment]
This turned out to be more work than its worth -- gcc-13 is sufficient.

**Comment by tchaikov** [comment]
just wanted to note, GCC-13 might suffer from some ICE when compiling coroutine. see https://github.com/scylladb/seastar/pull/2207 and https://gcc.gnu.org/bugzilla/show_bug.cgi?id=112557

**Pull request closed without merging**